autogen/agentchat/contrib/capabilities/transform_messages.py

```diff
@@ -0,0 +1,278 @@
+
+        This function performs the following modifications to the agent:
+
+        1. Registers a hook that automatically transforms all messages before they are processed for
+            response generation.
+        """
+        agent.register_hook(hookable_method="process_all_messages_before_reply", hook=self._transform_messages)
+
+    def _transform_messages(self, messages: List[Dict]) -> List[Dict]:
+        temp_messages = copy.deepcopy(messages)
+        system_message = None
+
+        if messages[0]["role"] == "system":
+            system_message = copy.deepcopy(messages[0])
+            temp_messages.pop(0)
+
+        for transform in self._transforms:
+            temp_messages = transform.apply_transform(temp_messages)
+
+        if system_message:
+            temp_messages.insert(0, system_message)
+
+        self._print_stats(messages, temp_messages)
+
+        return temp_messages
+
+    def _print_stats(self, pre_transform_messages: List[Dict], post_transform_messages: List[Dict]):
+        pre_transform_messages_len = len(pre_transform_messages)
+        post_transform_messages_len = len(post_transform_messages)
+
+        if pre_transform_messages_len < post_transform_messages_len:
+            print(
+                colored(
+                    f"Number of messages reduced from {pre_transform_messages_len} to {post_transform_messages_len}.",
+                    "yellow",
+                )
+            )
+
+
+class MessageHistoryLimiter:
+    """Limits the number of messages considered by an agent for response generation.
+
+    This transform keeps only the most recent messages up to the specified maximum number of messages (max_messages).
+    It trims the conversation history by removing older messages, retaining only the most recent messages.
+    """
+
+    def __init__(self, max_messages: Optional[int] = None):
+        """
+        Args:
+            max_messages (None or int): Maximum number of messages to keep in the context.
+            Must be greater than 0 if not None.
+        """
+        self._validate_max_messages(max_messages)
+        self._max_messages = max_messages
+
+    def apply_transform(self, messages: List[Dict]) -> List[Dict]:
+        """Truncates the conversation history to the specified maximum number of messages.
+
+        This method returns a new list containing the most recent messages up to the specified
+        maximum number of messages (max_messages). If max_messages is None, it returns the
+        original list of messages unmodified.
+
+        Args:
+            messages (List[Dict]): The list of messages representing the conversation history.
+
+        Returns:
+            List[Dict]: A new list containing the most recent messages up to the specified maximum.
+        """
+        if self._max_messages is None:
+            return messages
+
+        return messages[-self._max_messages :]
```

Comment:
Sometimes, it is better to remove the messages in the middle rather than the head, because often the first 1 or 2 messages contains the highest-level of task.

Should we add a parameter in the __init__ function, `truncate_strategy: str = "head"`, where it could be "head", "middle", etc.

Reply:
Interesting idea, I didn't think about this. What do you think about the idea of having a single transform per strategy? I'm also thinking about doing this to `MessageTokenLimiter`, splitting it up into `HistoryTokenLimiter` and `MessageTokenLimiter` (still didn't finalize the names).

Reply:
Also for the implementation, maybe the argument could `skip: Optional[List[int]] = None`, and if the user wants to skip the first two messages, he can do `skip=[0,1]`